crates/walrus-sdk/src/client/quilt_client.rs

```diff
@@ -0,0 +1,583 @@
+            .await?;
+        let metadata = self
+            .client
+            .retrieve_metadata(certified_epoch, quilt_id)
+            .await?;
+
+        // Try to retrieve the quilt index from the slivers.
+        let quilt_index =
+            if let Ok(quilt_index) = self.retrieve_quilt_index(&metadata, certified_epoch).await {
+                quilt_index
+            } else {
+                // If the quilt index cannot be retrieved from the slivers, try to retrieve the
+                // quilt.
+                tracing::debug!(
+                    "failed to retrieve index slivers, trying to get quilt instead {}",
+                    quilt_id
+                );
+                self.get_quilt_enum(&metadata, certified_epoch)
+                    .await?
+                    .get_quilt_index()?
+            };
+
+        let quilt_metadata = match quilt_index {
+            QuiltIndex::V1(quilt_index) => QuiltMetadata::V1(QuiltMetadataV1 {
+                quilt_blob_id: *quilt_id,
+                metadata: metadata.metadata().clone(),
+                index: quilt_index.clone(),
+            }),
+        };
+
+        Ok(quilt_metadata)
+    }
+
+    /// Retrieves the necessary slivers and decodes the quilt index.
+    ///
+    /// Returns error if not enough slivers can be retrieved for the quilt index.
+    async fn retrieve_quilt_index(
+        &self,
+        metadata: &VerifiedBlobMetadataWithId,
+        certified_epoch: Epoch,
+    ) -> ClientResult<QuiltIndex> {
+        // Get the first sliver to determine the quilt version.
+        //
+        // Since the quilt version is stored as the first byte of the Quilt, it doesn't matter
+        // whether we get the first primary sliver or the first secondary sliver.
+        // For now since we only support QuiltV1, we use the first secondary sliver.
+        let slivers = self
+            .client
+            .retrieve_slivers_with_retry::<Secondary>(
+                metadata,
+                &[SliverIndex::new(0)],
+                certified_epoch,
+                Self::MAX_RETRIEVE_SLIVERS_ATTEMPTS,
+                Self::TIMEOUT_DURATION,
+            )
+            .await?;
+
+        let first_sliver = slivers.first().expect("the first sliver should exist");
+        let quilt_version = QuiltVersionEnum::new_from_sliver(first_sliver.symbols.data())?;
+
+        let quilt_index = match quilt_version {
+            QuiltVersionEnum::V1 => {
+                self.retrieve_quilt_index_internal::<QuiltVersionV1>(
+                    metadata,
+                    certified_epoch,
+                    &slivers,
+                )
+                .await?
+            }
+        };
+
+        Ok(quilt_index)
+    }
+
+    async fn retrieve_quilt_index_internal<V: QuiltVersion>(
+        &self,
+        metadata: &VerifiedBlobMetadataWithId,
+        certified_epoch: Epoch,
+        first_sliver: &[SliverData<V::SliverAxis>],
```

Comment:
Nit: This doesn't need to be a slice.
```suggestion
        first_sliver: &SliverData<V::SliverAxis>,
```